### PR TITLE
OS X Handoff Feature

### PR DESCRIPTION
**Pull request by CharlieHess**
This PR implements OS X [Handoff](https://developer.apple.com/library/ios/documentation/UserExperience/Conceptual/Handoff/HandoffFundamentals/HandoffFundamentals.html) using new APIs on `app`, as follows:

```
/**
 * Creates an NSUserActivity and sets it as the current activity.
 * Other Mac and iOS apps can retrieve this data in an app delegate.
 */
app.setUserActivity('com.Electron.myEvent', {
  dataToResumeWith: 'abc',
  only: 'strings allowed'
});
```

```
/**
 * If your Info.plist includes this activity type, your app will receive this 
 * event when a "handoff" occurs.
 */
app.on("continue-activity", (e, type, userInfo) => {
  if (type === 'com.Electron.myEvent') {
    event.preventDefault();
    // resume activity using data from userInfo
  }
});
```
In its current form, you can only pass strings in the `userInfo` object. Someone with more Objective-C++ chops could certainly improve on this.

### TODO:
- [x] Documentation
- [ ] Code review

### CODE REVIEW

File: docs/api/app.md [hunk 1]
```diff
@@ -108,8 +108,20 @@ Returns:
...
+Emitted when the application is activated, which usually happens when the user clicks on
+the application's dock icon.
+
+### Event: 'continue-activity' _OS X_
+
+Returns:
+
+* `event` Event
+* `type` String - A string identifying the event. Maps to [`NSUserActivity.activityType`](https://developer.apple.com/library/ios/documentation/Foundation/Reference/NSUserActivity_Class/index.html#//apple_ref/occ/instp/NSUserActivity/activityType).
+* `userInfo` Object - Contains app-specific state stored by the activity on another device.
```
Comment (paulcbetts) [inline docs/api/app.md]: Does this userInfo object have to be a dictionary of strings?

**Comment by dasilvacontin** [comment]
Thanks so much for working on this!

File: atom/browser/browser_mac.mm [hunk 2]
```diff
@@ -87,6 +87,23 @@
...
+  NSString* type_ns = [NSString stringWithUTF8String:type.c_str()];
+  NSUserActivity *user_activity = [[NSUserActivity alloc] initWithActivityType:type_ns];
+
+  NSMutableArray* user_info_args = [[NSMutableArray alloc] init];
+  for (auto const &pair : user_info) {
+    NSString* key_ns = [NSString stringWithUTF8String:pair.first.c_str()];
+    NSString* value_ns = [NSString stringWithUTF8String:pair.second.c_str()];
+    
+    [user_info_args addObject:key_ns];
+    [user_info_args addObject:value_ns];
```
Comment (dasilvacontin) [inline atom/browser/browser_mac.mm]: I believe the order is value,key,value,key,value,key etc https://developer.apple.com/library/mac/documentation/Cocoa/Reference/Foundation/Classes/NSDictionary_Class/index.html#//apple_ref/occ/instm/NSDictionary/initWithObjectsAndKeys:

File: atom/browser/mac/mac_native_converter.mm [hunk 3]
```diff
@@ -0,0 +1,54 @@
+#import "atom/browser/mac/mac_native_converter.h"
+
+@implementation MacNativeConverter
```
Comment (CharlieHess) [inline atom/browser/mac/mac_native_converter.mm]: This is getting a little out of hand and I feel like I must be duplicating code that already exists somewhere.
@zcbenz: Is there a better way to convert between Cocoa Foundation types and `base` types?

File: docs/api/app.md (hunk 1 above)
Comment (CharlieHess) [inline docs/api/app.md]: Nope, that was a dumb simplification that I forgot about. I changed it to pass around `base::Dictionary`; only issue is that converting back and forth (taking nesting into account) is a lot of boilerplate.

Comment (paulcbetts) [inline docs/api/app.md]: I mean, I actually think it's quite reasonable to constrain it to strings, it'd probably make your code way simpler

File: atom/browser/browser_mac.mm [hunk 4]
```diff
@@ -87,6 +87,23 @@
 void Browser::SetAppUserModelID(const base::string16& name) {
 }
 
+void Browser::SetUserActivity(const std::string& type, const std::map<std::string, std::string>& user_info) {
+  NSString* type_ns = [NSString stringWithUTF8String:type.c_str()];
+  NSUserActivity *user_activity = [[NSUserActivity alloc] initWithActivityType:type_ns];
```
Comment (zcbenz) [inline atom/browser/browser_mac.mm]: The `user_activity`, `user_info_args`, and the array passed to `user_activity.userInfo` should be released after being used, you can put them in a `base::scoped_nsobject` to make them release automatically.

**Comment by zcbenz** [comment]
👍

**Pull request merged**